main.go

```diff
@@ -14,7 +14,312 @@
+
+	tmpDir, err := ioutil.TempDir("", config.Name)
+	if err != nil {
+		logrus.Fatalf("error creating tmpDir %+v", err)
+	}
+	defer func(tmpDir string) { _ = os.Remove(tmpDir) }(tmpDir)
+	listenOn := &(url.URL{Scheme: "unix", Path: filepath.Join(tmpDir, config.ListenOn)})
+
+	srvErrCh := grpcutils.ListenAndServe(ctx, listenOn, server)
+	exitOnErr(ctx, cancel, srvErrCh)
+	log.FromContext(ctx).Infof("grpc server started")
+
+	// ********************************************************************************
+	log.FromContext(ctx).Infof("executing phase 6: register nse with nsm")
+	// ********************************************************************************
+
+	nseRegistryClient := registryclient.NewNetworkServiceEndpointRegistryClient(ctx, &config.ConnectTo, registryclient.WithDialOptions(clientOptions...))
+	nse, err := nseRegistryClient.Register(ctx, &registryapi.NetworkServiceEndpoint{
+		Name:                config.Name,
+		NetworkServiceNames: []string{config.ServiceName},
+		NetworkServiceLabels: map[string]*registryapi.NetworkServiceLabels{
+			config.ServiceName: {
+				Labels: config.Labels,
+			},
+		},
+		Url: listenOn.String(),
+	})
+	logrus.Infof("nse: %+v", nse)
+
+	if err != nil {
+		log.FromContext(ctx).Fatalf("unable to register nse %+v", err)
+	}
+
+	// ********************************************************************************
+	log.FromContext(ctx).Infof("startup completed in %v", time.Since(starttime))
+	// ********************************************************************************
+
+	// wait for server to exit
+	<-ctx.Done()
+	<-vppErrCh
+}
+
+func exitOnErr(ctx context.Context, cancel context.CancelFunc, errCh <-chan error) {
+	// If we already have an error, log it and exit
+	select {
+	case err := <-errCh:
+		log.FromContext(ctx).Fatal(err)
+	default:
+	}
+	// Otherwise wait for an error in the background to log and cancel
+	go func(ctx context.Context, errCh <-chan error) {
+		err := <-errCh
+		log.FromContext(ctx).Error(err)
+		cancel()
+	}(ctx, errCh)
+}
+
+func notifyContext() (context.Context, context.CancelFunc) {
+	return signal.NotifyContext(
+		context.Background(),
+		os.Interrupt,
+		// More Linux signals here
+		syscall.SIGHUP,
+		syscall.SIGTERM,
+		syscall.SIGQUIT,
+	)
+}
+
+func logPhases(ctx context.Context) {
+	// enumerating phases
+	log.FromContext(ctx).Infof("there are 6 phases which will be executed followed by a success message:")
+	log.FromContext(ctx).Infof("the phases include:")
+	log.FromContext(ctx).Infof("1: get config from environment")
+	log.FromContext(ctx).Infof("2: retrieve spiffe svid")
+	log.FromContext(ctx).Infof("3: create grpc client options")
+	log.FromContext(ctx).Infof("4: create firewall network service endpoint")
+	log.FromContext(ctx).Infof("5: create grpc and mount nse")
+	log.FromContext(ctx).Infof("6: register nse with nsm")
+	log.FromContext(ctx).Infof("a final success message with start time duration")
+}
+
+func setupLogging(ctx context.Context) {
```

Comment:
this function is incorrect...
We should use context with logger in main function.

Please take a look at what is structure data context: https://golang.org/pkg/context/#WithValue

Reply:
ok, fixed